### DISCUSSION TITLE
[SWIP-5][DISCUSSION] Support ClickHouse monitoring

### DISCUSSION THREAD

**Discussion by CzyerChen** [apache/skywalking#11904] [SWIP]
## Motivation

ClickHouse is a high-performance, column-oriented SQL database management system (DBMS) for online analytical processing (OLAP). It is available as both an [open-source software](https://github.com/ClickHouse/ClickHouse) and a [cloud offering](https://clickhouse.com/cloud). Now I want to add ClickHouse monitoring via the OpenTelemetry Collector, which fetches metrics from it's own HTTP endpoint to expose metrics data for [Prometheus](https://prometheus.io/)(since ClickHouse v20.1.2.4). Clickhouse Exporter used only for old ClickHouse versions, modern versions have embedded prometheus endpoint.

## Architecture Graph

There is no significant architecture-level change.

## Proposed Changes

`ClickHouse`  expose own metrics via  HTTP endpoint to opentelemetry collector,using skyWalking openTelemetry receiver to receive these metrics.
Expose metrics from the [system.metrics](https://clickhouse.com/docs/en/operations/system-tables/metrics#system_tables-metrics) table / the [system.events](https://clickhouse.com/docs/en/operations/system-tables/events#system_tables-events) table / the [system.asynchronous_metrics](https://clickhouse.com/docs/en/operations/system-tables/asynchronous_metrics#system_tables-asynchronous_metrics) table.

### ClickHouse Instance Supported Metrics

| Monitoring Panel | Unit       | Metric Name                                | Description                                                                                                      | Data Source |
| ---------------- | ---------- | ------------------------------------------ | ---------------------------------------------------------------------------------------------------------------- | ----------- |
| CpuUsage         | count      | meter_clickhouse_instance_cpu_usage        | CPU time spent seen by OS per second(according to ClickHouse.system.dashboard.CPU Usage (cores)).                | ClickHouse  |
| MemoryUsage      | percentage | meter_clickhouse_instance_memory_usage     | Total amount of memory (bytes) allocated by the server/ total amount of OS memory.                               | ClickHouse  |
| MemoryAvailable  | percentage | meter_clickhouse_instance_memory_available | Total amount of memory (bytes) available for program / total amount of OS memory.                                | ClickHouse  |
| Uptime           | sec        | meter_clickhouse_instance_uptime           | The server uptime in seconds. It includes the time spent for server initialization before accepting connections. | ClickHouse  |
| Version          | string     | meter_clickhouse_instance_version          | Version of the server in a single integer number in base-1000.                                                   | ClickHouse  |
| FileOpen         | count      | meter_clickhouse_instance_file_open        | Number of files opened.                                                                                          | ClickHouse  |

### ClickHouse Network Supported Metrics

| Monitoring Panel       | Unit  | Metric Name                                                                                      | Description                                                | Data Source |
| ---------------------- | ----- | ------------------------------------------------------------------------------------------------ | ---------------------------------------------------------- | ----------- |
| TcpConnections         | count | meter_clickhouse_instance_tcp_connections<br/>meter_clickhouse_tcp_connections                   | Number of connections to TCP server.                       | ClickHouse  |
| MysqlConnections       | count | meter_clickhouse_instance_mysql_connections<br/>meter_clickhouse_mysql_connections               | Number of client connections using MySQL protocol.         | ClickHouse  |
| HttpConnections        | count | meter_clickhouse_instance_http_connections<br/>meter_clickhouse_mysql_connections                | Number of connections to HTTP server.                      | ClickHouse  |
| InterserverConnections | count | meter_clickhouse_instance_interserver_connections<br/>meter_clickhouse_interserver_connections   | Number of connections from other replicas to fetch parts.  | ClickHouse  |
| PostgresqlConnections  | count | meter_clickhouse_instance_postgresql_connections<br/>meter_clickhouse_postgresql_connections     | Number of client connections using PostgreSQL protocol.    | ClickHouse  |
| ReceiveBytes           | bytes | meter_clickhouse_instance_network_receive_bytes<br/>meter_clickhouse_network_receive_bytes       | Total number of bytes received from network.               | ClickHouse  |
| SendBytes              | bytes | meter_clickhouse_instance_network_send_bytes<br/>meter_clickhouse_network_send_bytes             | Total number of bytes send to network.                     | ClickHouse  |

### ClickHouse Query Supported Metrics

| Monitoring Panel | Unit      | Metric Name                                                                                                 | Description                                               | Data Source |
| ---------------- | --------- | ----------------------------------------------------------------------------------------------------------- | --------------------------------------------------------- | ----------- |
| QueryCount       | count     | meter_clickhouse_instance_query<br/>meter_clickhouse_query                                                  | Number of executing queries.                              | ClickHouse  |
| SelectQueryCount | count     | meter_clickhouse_instance_query_select<br/>meter_clickhouse_query_select                                    | Number of executing queries, but only for SELECT queries. | ClickHouse  |
| InsertQueryCount | count     | meter_clickhouse_instance_query_insert<br/>meter_clickhouse_query_insert                                    | Number of executing queries, but only for INSERT queries. | ClickHouse  |
| SelectQueryRate  | count/sec | meter_clickhouse_instance_query_select_rate<br/>meter_clickhouse_query_select_rate                          | Number of SELECT queries per second.                      | ClickHouse  |
| InsertQueryRate  | count/sec | meter_clickhouse_instance_query_insert_rate<br/>meter_clickhouse_query_insert_rate                          | Number of INSERT queries per second.                      | ClickHouse  |
| Querytime        | microsec  | meter_clickhouse_instance_querytime_microseconds<br/>meter_clickhouse_querytime_microseconds                | Total time of all queries.                                | ClickHouse  |
| SelectQuerytime  | microsec  | meter_clickhouse_instance_querytime_select_microseconds<br/>meter_clickhouse_querytime_select_microseconds  | Total time of SELECT queries.                             | ClickHouse  |
| InsertQuerytime  | microsec  | meter_clickhouse_instance_querytime_insert_microseconds<br/>meter_clickhouse_querytime_insert_microseconds  | Total time of INSERT queries.                             | ClickHouse  |
| OtherQuerytime   | microsec  | meter_clickhouse_instance_querytime_other_microseconds<br/>meter_clickhouse_querytime_other_microseconds    | Total time of queries that are not SELECT or INSERT.      | ClickHouse  |
| QuerySlowCount   | count     | meter_clickhouse_instance_query_slow<br/>meter_clickhouse_query_slow                                        | Number of reads from a file that were slow.               | ClickHouse  |

### ClickHouse Insertion Supported Metrics

| Monitoring Panel   | Unit  | Metric Name                                                                  | Description                                                                                                                     | Data Source |
| ------------------ | ----- | ---------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| InsertQueryCount   | count | meter_clickhouse_instance_query_insert<br/>meter_clickhouse_query_insert     | Number of executing queries, but only for INSERT queries.                                                                       | ClickHouse  |
| InsertedRowCount   | count | meter_clickhouse_instance_inserted_rows<br/>meter_clickhouse_inserted_rows   | Number of rows INSERTed to all tables.                                                                                          | ClickHouse  |
| InsertedBytes      | bytes | meter_clickhouse_instance_inserted_bytes<br/>meter_clickhouse_inserted_bytes | Number of bytes INSERTed to all tables.                                                                                         | ClickHouse  |
| DelayedInsertCount | count | meter_clickhouse_instance_delayed_insert<br/>meter_clickhouse_delayed_insert | Number of times the INSERT of a block to a MergeTree table was throttled due to high number of active data parts for partition. | ClickHouse  |

### ClickHouse Replicas Supported Metrics

| Monitoring Panel | Unit  | Metric Name                                                  | Description                                      | Data Source |
| ---------------- | ----- | ------------------------------------------------------------ | ------------------------------------------------ | ----------- |
| ReplicatedChecks | count | meter_clickhouse_instance_replicated_checks<br/>meter_clickhouse_replicated_checks | Number of data parts checking for consistency.   | ClickHouse  |
| ReplicatedFetch  | count | meter_clickhouse_instance_replicated_fetch<br/>meter_clickhouse_replicated_fetch | Number of data parts being fetched from replica. | ClickHouse  |
| ReplicatedSend   | count | meter_clickhouse_instance_replicated_send<br/>meter_clickhouse_replicated_send | Number of data parts being sent to replicas.     | ClickHouse  |

### ClickHouse MergeTree Supported Metrics

| Monitoring Panel       | Unit  | Metric Name                                                                                      | Description                                                                                                                     | Data Source |
| ---------------------- | ----- | ------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| BackgroundMergeCount   | count | meter_clickhouse_instance_background_merge<br/>meter_clickhouse_background_merge                 | Number of executing background merges.                                                                                          | ClickHouse  |
| MergeRows              | count | meter_clickhouse_instance_merge_rows<br/>meter_clickhouse_merge_rows                             | Rows read for background merges. This is the number of rows before merge.                                                       | ClickHouse  |
| MergeUncompressedBytes | bytes | meter_clickhouse_instance_merge_uncompressed_bytes<br/>meter_clickhouse_merge_uncompressed_bytes | Uncompressed bytes (for columns as they stored in memory) that was read for background merges. This is the number before merge. | ClickHouse  |
| MoveCount              | count | meter_clickhouse_instance_move<br/>meter_clickhouse_move                                         | Number of currently executing moves.                                                                                            | ClickHouse  |
| PartsActive            | Count | meter_clickhouse_instance_parts_active<br/>meter_clickhouse_parts_active                         | Active data part, used by current and upcoming SELECTs.                                                                         | ClickHouse  |
| MutationsCount         | count | meter_clickhouse_instance_mutations<br/>meter_clickhouse_mutations                               | Number of mutations (ALTER DELETE/UPDATE).                                                                                      | ClickHouse  |

### ClickHouse Kafka Table Engine Supported Metrics

When [table engine](https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka) works
with [Apache Kafka](http://kafka.apache.org/).

Kafka lets you:

- Publish or subscribe to data flows.
- Organize fault-tolerant storage.
- Process streams as they become available.

| Monitoring Panel  | Unit  | Metric Name                                                                            | Description                                               | Data Source |
| ----------------- | ----- | -------------------------------------------------------------------------------------- | --------------------------------------------------------- | ----------- |
| KafkaMessagesRead | count | meter_clickhouse_instance_kafka_messages_read<br/>meter_clickhouse_kafka_messages_read | Number of Kafka messages already processed by ClickHouse. | ClickHouse  |
| KafkaWrites       | count | meter_clickhouse_instance_kafka_writes<br/>meter_clickhouse_kafka_writes               | Number of writes (inserts) to Kafka tables.               | ClickHouse  |
| KafkaConsumers    | count | meter_clickhouse_instance_kafka_consumers<br/>meter_clickhouse_kafka_consumers         | Number of active Kafka consumers.                         | ClickHouse  |
| KafkProducers     | count | meter_clickhouse_instance_kafka_producers<br/>meter_clickhouse_kafka_producers         | Number of active Kafka producer created.                  | ClickHouse  |

### ClickHouse ZooKeeper Supported Metrics

ClickHouse uses ZooKeeper for storing metadata of replicas when using replicated tables. If replicated tables are not
used, this section of parameters can be omitted.

| Monitoring Panel      | Unit  | Metric Name                                                                                       | Description                                                           | Data Source |
| --------------------- | ----- | ------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------- | ----------- |
| ZookeeperSession      | count | meter_clickhouse_instance_zookeeper_session<br/>meter_clickhouse_zookeeper_session                | Number of sessions (connections) to ZooKeeper.                        | ClickHouse  |
| ZookeeperWatch        | count | meter_clickhouse_instance_zookeeper_watch<br/>meter_clickhouse_zookeeper_watch                    | Number of watches (event subscriptions) in ZooKeeper.                 | ClickHouse  |
| ZookeeperBytesSent    | bytes | meter_clickhouse_instance_zookeeper_bytes_sent<br/>meter_clickhouse_zookeeper_bytes_sent          | Number of bytes send over network while communicating with ZooKeeper. | ClickHouse  |
| ZookeeperBytesReceive | bytes | meter_clickhouse_instance_zookeeper_bytes_received<br/>meter_clickhouse_zookeeper_bytes_received  | Number of bytes send over network while communicating with ZooKeeper. | ClickHouse  |

### ClickHouse Keeper Supported Metrics

| Monitoring Panel         | Unit  | Metric Name                                                                                           | Description                                                              | Data Source |
| ------------------------ | ----- | ----------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------ | ----------- |
| KeeperAliveConnections   | count | meter_clickhouse_instance_keeper_connections_alive<br/>meter_clickhouse_keeper_connections_alive      | Number of alive connections for embedded ClickHouse Keeper.              | ClickHouse  |
| KeeperOutstandingRequets | count | meter_clickhouse_instance_keeper_outstanding_requests<br/>meter_clickhouse_keeper_outstanding_requests| Number of outstanding requests for embedded ClickHouse Keeper.           | ClickHouse  |

## Imported Dependencies libs and their licenses.

No new dependency.

## Compatibility

no breaking changes.

**Comment by wu-sheng** [comment]
Do you consider the `service hierarchy` between this and virtual DB? I checked the plugin supported list, at least, clickhouse-JDBC has been [supported in Java agent](https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/supported-list/).

**Comment by wu-sheng** [comment]
Could you update the SWIP about how would you catalog these metrics into the standard service/instance/endpoint concept? The metrics you listed are from the original system, but you haven't indicated how these would look like through MAL and eventually visualized in SkyWalking concepts.

**Reply by wu-sheng** [reply]
@CzyerChen I think you missed this in this update. We should write these in the SWIP.

**Reply by CzyerChen** [reply]
### Data flow
1. Configure ClickHouse to expose metrics for Prometheus Receiver.
2. OpenTelemetry Collector fetches metrics from Prometheus Receiver and pushes metrics to SkyWalking OAP Server via OpenTelemetry gRPC exporter.
3. The SkyWalking OAP Server parses the expression with [MAL](../../concepts-and-designs/mal.md) to filter/calculate/aggregate and store the results.

### Set up
1. Set up [built-in prometheus endpoint](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#server_configuration_parameters-prometheus).

```    
    <prometheus>
        <endpoint>/metrics</endpoint>
        <port>9363</port>
        <metrics>true</metrics>
        <events>true</events>
        <asynchronous_metrics>true</asynchronous_metrics>
    </prometheus>
```

2. Set up [OpenTelemetry Collector ](https://opentelemetry.io/docs/collector/getting-started/#docker). For details on Prometheus Receiver in OpenTelemetry Collector
```
version: '3.8'
services: 
  otel-collector:
    image: otel/opentelemetry-collector:latest
    container_name: otel-collector
    network_mode: host
    command: [ "--config=/etc/otel-collector-config.yaml" ]
    volumes:
      - ./otel-collector-config.yaml:/etc/otel-collector-config.yaml
    expose:
      - 5567
```


3. Config SkyWalking [OpenTelemetry receiver](opentelemetry-receiver.md).
```
receivers:
  prometheus:
    config:
     scrape_configs:
       - job_name: 'clickhouse-monitoring'
         scrape_interval: 15s
         static_configs:
           - targets: ['ip1:9363','1ip2:9363','ip3:9363']
             labels:
               host_name: prometheus-clickhouse
processors:
  batch:
 
exporters:
  otlp:
    endpoint: ip:11800
    tls:
      insecure: true
service:
  pipelines:
    metrics:
      receivers:
      - prometheus
      processors:
      - batch
      exporters:
      - otlp
```

**Reply by wu-sheng** [reply]
> how would you catalog these metrics into the standard service/instance/endpoint concept? 

Could you add definition of these mapping in the SWIP?

**Comment by wu-sheng** [comment]
Any update here? I noticed this SWIP has no update for a week.

**Reply by CzyerChen** [reply]
I will update it today with some changes of meter and  add some images.

**Comment by CzyerChen** [comment]
1. Add CPU/Memory Usage into Instance.
CPU Usage according to the calculation of official ClickHouse [system.dashboard](https://clickhouse.com/docs/en/operations/system-tables/dashboards)
3. Simple view of the dashboard about the ClickHouse cluster with 1 ZooKeeper Node and 2 ClickHouse DB Node.
<img width="1428" alt="Skywalking" src="https://github.com/apache/skywalking/assets/22817918/3c16862b-1b6c-42e8-a993-baa1c538ea88">
<img width="1436" alt="Pasted Graphic 1" src="https://github.com/apache/skywalking/assets/22817918/7db132d8-f268-4afc-a428-faa066f4f4ca">
<img width="1434" alt="Pasted Graphic 2" src="https://github.com/apache/skywalking/assets/22817918/6a7345ac-a154-4e70-9653-fab0e23c3dda">
<img width="1431" alt="Pasted Graphic 3" src="https://github.com/apache/skywalking/assets/22817918/36963dfe-d125-48cd-a2d9-48c62d68bddf">
<img width="1434" alt="ClickHouse-Root" src="https://github.com/apache/skywalking/assets/22817918/9d53f58e-a607-4fd3-bb9f-914b1872574c">
<img width="1428" alt="Pasted Graphic 5" src="https://github.com/apache/skywalking/assets/22817918/f73273d4-b13a-4803-9a11-daf7c165d2e6">
<img width="1440" alt="ClickHouse-Root ClickHouse-Service  ClickHouse-Instance" src="https://github.com/apache/skywalking/assets/22817918/f37cbf13-2da4-431f-b630-543bf9854fcf">

I have to check the metrics value again.

**Reply by CzyerChen** [reply]
I also will test `service hierarchy` concept later.

**Comment by CzyerChen** [comment]
I have encountered two problems during working on the dashboard that can not solved.
1. ClickHouse version string give in base-1000 format like `version 11.22.33 is translated to 11022033`, is it possible to translate 11022033 into  11.22.33 when collecting metrics?
2. When using table format in dashboard with two meter metrics and refreshing data with icon at top right, the table data sometimes will not be updated. Can it be optimized when data reloading?
<img width="696" alt="Memory()" src="https://github.com/apache/skywalking/assets/22817918/dad38b16-c200-48e2-a843-7b19f39e61d4">

so do the demo in mongo monitoring.

<img width="1363" alt="DB Avg Data Size Per Shard (GB)" src="https://github.com/apache/skywalking/assets/22817918/8e5920f9-5a17-427b-a001-225e3f5b4d3a">

**Reply by wu-sheng** [reply]
> ClickHouse version string give in base-1000 format like version 11.22.33 is translated to 11022033, is it possible to translate 11022033 into 11.22.33 when collecting metrics?

Where do you get that? Which type does OTLP provide for this?

> When using table format in dashboard with two meter metrics and refreshing data with icon at top right, the table data sometimes will not be updated. Can it be optimized when data reloading?

Do you have a temp env to make @Fine0830 test and verify?

**Reply by wu-sheng** [reply]
@Fine0830 Could you check whether this widget has similar issues with https://github.com/apache/skywalking-booster-ui/pull/373?

**Reply by CzyerChen** [reply]
Just from `system.metrics` in ClickHouse

```

iZ2ze72qvclbhzjoaz5c2jZ :) select * from system.metrics where metric ='VersionInteger' limit 1;

SELECT *
FROM system.metrics
WHERE metric = 'VersionInteger'
LIMIT 1

Query id: b0189d45-9054-4a01-9506-586d216661ed

┌─metric─────────┬────value─┬─description─────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ VersionInteger │ 23012002 │ Version of the server in a single integer number in base-1000. For example, version 11.22.33 is translated to 11022033. │
└────────────────┴──────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.007 sec. 
```

If possible, we can show the version data.

```
# HELP ClickHouseMetrics_VersionInteger Version of the server in a single integer number in base-1000. For example, version 11.22.33 is translated to 11022033.
# TYPE ClickHouseMetrics_VersionInteger gauge
ClickHouseMetrics_VersionInteger 23012002
```

**Reply by wu-sheng** [reply]
Show `23012002` is fine. But it may be hard to show as `23.12.2`. @wankai123 Any suggestion?

**Reply by wankai123** [reply]
I think show `23012002` and adding some tips is enough.

**Comment by wu-sheng** [comment]
PR continues on https://github.com/apache/skywalking/pull/11966. This is accepted.

**Comment by CzyerChen** [comment]
Final version could be found https://skywalking.apache.org/docs/main/next/en/swip/swip-5/
